streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyStore.java

```diff
@@ -0,0 +1,136 @@
+ * copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.QueryableStoreType;
+import org.apache.kafka.streams.state.ReadOnlyKeyValueStore;
+import org.apache.kafka.streams.state.StateStoreProvider;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+/**
+ * A wrapper over the underlying {@link ReadOnlyKeyValueStore}s found in a {@link
+ * org.apache.kafka.streams.processor.internals.ProcessorTopology}
+ *
+ * @param <K> key type
+ * @param <V> value type
+ */
+public class CompositeReadOnlyStore<K, V> implements ReadOnlyKeyValueStore<K, V> {
+
+    private final StateStoreProvider storeProvider;
+    private final QueryableStoreType<ReadOnlyKeyValueStore<K, V>> storeType;
+    private final String storeName;
+
+    public CompositeReadOnlyStore(final StateStoreProvider storeProvider,
+                                  final QueryableStoreType<ReadOnlyKeyValueStore<K, V>> storeType,
+                                  final String storeName) {
+        this.storeProvider = storeProvider;
+        this.storeType = storeType;
+        this.storeName = storeName;
+    }
+
+    @Override
+    public V get(final K key) {
+        final List<ReadOnlyKeyValueStore<K, V>> stores = storeProvider.getStores(storeName, storeType);
+        for (ReadOnlyKeyValueStore<K, V> store : stores) {
+            V result = store.get(key);
+            if (result != null) {
+                return result;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public KeyValueIterator<K, V> range(final K from, final K to) {
+        final NextIteratorFunction<K, V> nextIteratorFunction = new NextIteratorFunction<K, V>() {
+            @Override
+            public KeyValueIterator<K, V> apply(final ReadOnlyKeyValueStore<K, V> store) {
+                return store.range(from, to);
+            }
+        };
+        final List<ReadOnlyKeyValueStore<K, V>> stores = storeProvider.getStores(storeName, storeType);
+        return new CompositeKeyValueIterator(stores.iterator(), nextIteratorFunction);
+    }
+
+    @Override
+    public KeyValueIterator<K, V> all() {
+        final NextIteratorFunction<K, V> nextIteratorFunction = new NextIteratorFunction<K, V>() {
+            @Override
+            public KeyValueIterator<K, V> apply(final ReadOnlyKeyValueStore<K, V> store) {
+                return store.all();
```

Comment:
And looking into it further, there is an additional check in the `RocksDBStore` when `all` or `range` is called. if the store is not `open` an exception will be thrown - which i think is valid in this case. It would mean a re-balance is in process.